py_snappy/main.py

```diff
@@ -0,0 +1,342 @@
+#
+# For copy tags, length bytes are copied from offset bytes ago, in the style of
+# Lempel-Ziv compression algorithms. In particular:
+#   - For l == 1, the offset ranges in [0, 1<<11) and the length in [4, 12).
+#     The length is 4 + the low 3 bits of m. The high 3 bits of m form bits
+#     8-10 of the offset. The next byte is bits 0-7 of the offset.
+#   - For l == 2, the offset ranges in [0, 1<<16) and the length in [1, 65).
+#     The length is 1 + m. The offset is the little-endian unsigned integer
+#     denoted by the next 2 bytes.
+#   - For l == 3, this tag is a legacy format that is no longer supported.
+
+
+def uint8(n: int) -> int:
+    return n & ((1 << 8) - 1)
+
+
+def uint32(n: int) -> int:
+    return n & ((1 << 32) - 1)
+
+
+def uint64(n: int) -> int:
+    return n & ((1 << 64) - 1)
+
+
+def uvarint(buf: bytes) -> Tuple[int, int]:
+    """
+    uvarint decodes a uint64 from buf and returns that value and the number of
+    bytes read (> 0). If an error occurred, the value is 0 and the number of
+    bytes n is <= 0 meaning:
+    n == 0: buf too small
+    n  < 0: value larger than 64 bits (overflow)
+            and -n is the number of bytes read"""
+    x, s = 0, 0
+    for idx, b in enumerate(buf):
+        if b < 0x80:
+            if idx > 9 or (idx == 9 and b > 1):
+                return 0, -1 * (idx + 1)  # overflow
+            return x | uint64(b) << s, idx + 1
+        x |= uint64(b & 0x7F) << s
+        s += 7
+    return 0, 0
+
+
+TReturn = TypeVar("TReturn")
+
+
+def bytes_gen(fn: Callable[..., Iterable[int]]) -> Callable[..., bytes]:
+    @functools.wraps(fn)
+    def inner(*args: Any, **kwargs: Any) -> bytes:
+        return bytes(fn(*args, **kwargs))
+
+    return inner
+
+
+def tuple_gen(
+    fn: Callable[..., Iterable[TReturn]]
+) -> Callable[..., Tuple[TReturn, ...]]:
+    @functools.wraps(fn)
+    def inner(*args: Any, **kwargs: Any) -> Iterable[TReturn]:
+        return tuple(fn(*args, **kwargs))
+
+    return inner
+
+
+@bytes_gen
+def putuvarint(x: int) -> Iterable[int]:
+    """
+    putuvarint encodes a uint64.
+    """
+    while x >= 0x80:
+        yield uint8(x) | 0x80
+        x >>= 7
+    yield x
+
+
+def extract_meta(src: bytes) -> Tuple[int, int]:
+    """
+    Return a 2-tuple:
+
+    - the length of the decoded block
+    - the number of bytes that the length header occupied.
+  """
+    v, n = uvarint(src)
+    if n <= 0 or v > 0xFFFFFFFF:
+        raise CorruptError
+    if v > 0x7FFFFFFF:
+        raise TooLargeError
+    return v, n
+
+
+def decompress(buf: bytes) -> bytes:
+    """
+    decompress returns the decompressed form of buf.
+    """
+    block_length, s = extract_meta(buf)
+    src = tuple(c for c in buf)
+    src_len = len(src)
+    dst = [0] * block_length
+    d, offset, length = 0, 0, 0
+
+    while s < src_len:
+        b = src[s] & 0x03
+        if b == TAG_LITERAL:
+            x = src[s] >> 2
+
+            if x < 60:
+                s += 1
```

Comment:
Lots of single-letter variables: `d`, `b`, `s`, `x` makes it hard to keep track of what's going on. If they are referencing something specific to snappy or LZ compression, then I guess it's fine to leave. But if you have better intuition for reasonable names, then it would really help.

Reply:
I did a first pass expanding some of the names, I'll probably try to finish expanding them out in a subsequent PR.